src/hotspot/share/opto/gcm.cpp

```diff
@@ -608,6 +645,12 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
   Node_List worklist_store(area);   // possible-def to explore
   Node_List worklist_visited(area); // visited mergemem nodes
   Node_List non_early_stores(area); // all relevant stores outside of early
+
+  // Cases falling under unrelated_load_in_store_null_block can make
+  // raise_LCA_above_marks in inconsistent state, so always reset the
+  // visited blocks after use.
+  Block_List raise_LCA_visited_list;
```

Comment:
I'm not an expert, but this sounds like a risky change to me.  Could you explain the problem in more detail, and why resetting these values is safe?

Reply:
AFAIU reinit is ok. it is not reinitialized as it holds unique value.
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/block.hpp#L214

Reply:
I took another look, and it looks safe, but I'm concerned that it will cause raise_LCA_above_marks to do more work.  Don't you think there might be a performance impact?  Have you measured performance?

I still don't understand what inconsistency state this tries to solve, or how it solves it.

Reply:
raise_LCA_above_marks checks and marks visited bits from LCA and  move upward. i am trying to be safe by clearing visited bits where we don't use calculated LCA. marking visited and not used cases only get impacted.
Only performance issue is explicit clearing logic i guess.

Inconsistency  is marking visited and not using the newly computed LCA. but yes inconsistency can never happen as all the cases are computed by this time. and new store interfering  the loads can seldom come. but i was trying to be safe.

Reply:
I will remove the logic. as it is not anticipated case.

Reply:
> Only performance issue is explicit clearing logic i guess

The performance impact I had in mind was this short-circuit in raise_LCA_above_marks():

465      // Test and set the visited bit.
466      if (mid->raise_LCA_visited() == mark)  continue;  // already visited

If the visited mark is cleared, then raise_LCA_above_marks() needs to perform its work again.
Perhaps there is an assert that can guard against the inconsistency you were concerned about.